FreeTubeApp/FreeTube
# Implement right-to-left interface

# Implement right-to-left interface

Co-authored by @ChunkyProgrammer



## Pull Request Type


## Related issue
closes #1550

## Description

At a high level, this PR updates the interface of FreeTube to accommodate the proper patterns for right-to-left languages throughout the entire app. This PR looks quite large, and it is, but >80-85% of the changes are migrating CSS properties to their 1:1 directionality-agnostic equivalents.

Implementation-wise:
- **Doesn't touch VideoJS.css**
- Replaces `margin-left` / `-right` with `margin-inline-start` / `-end` & `margin-top` / `-bottom` with `margin-block-start` / `-end`
- Replaces `padding-left` / `-right` with `padding-inline-start` / `end` & `padding-top` / `-bottom` with `padding-block-start` / `-end`
- Replaces `left`/`right` with `inset-inline-start`/`-end` & `top`/`bottom` with `inset-block-start`/`-end`
- Replaces `border-left*`/`border-right*` with `border-inset-start`/`-end` & `border-top*`/`border-bottom*` with `border-block-start`/`-end`
- Replaces `text-align: left`/`right` with `text-align: start`/`end`
- Custom: replaces `float: left`/`right` with two CSS variables (because [`float: inline-start`/`-end` is still experimental in Chrome](https://developer.mozilla.org/en-US/docs/Web/CSS/float))
- Custom: modifies `transform()` x values to conditionally be inverted by multiplying them with a CSS variable of 1 or -1 based on language directionality (090f392)
- Custom styling was needed/added in one place (`ft-input.css` to fix an emergent issue with the search bar `magnifying-glass` icon positioning on languages with right-to-left directionality
- Fixes settings switches not being aligned when the locale has right-to-left directionality
- Updates arrow keys when the current locale has right-to-left directionality such that left arrow is the "history forward" arrow, and right arrow is the "history backward" arrow
- Refactors the arrow key logic to use class binding to resolve emergent bug with history arrows showing as disabled when changing the locale to one with different directionality. It's also better practice than whatever the original implementer was doing (me, in 2021, to be clear)

## Screenshots 

![Screenshot_20230829_022443](https://github.com/FreeTubeApp/FreeTube/assets/84899178/644aef02-97b8-459c-8073-99177a26c8ea)
![Screenshot_20230829_022536](https://github.com/FreeTubeApp/FreeTube/assets/84899178/e6288a77-d17a-4715-97ea-622086a1ee1d)
![Screenshot_20230829_022631](https://github.com/FreeTubeApp/FreeTube/assets/84899178/b55b40eb-a1cb-44cc-808d-c7a9eb2e8319)
![Screenshot_20230829_022711](https://github.com/FreeTubeApp/FreeTube/assets/84899178/a51b0d0b-1681-4d36-8fb5-bc596718828a)



## Testing 

### Test Plan
- Under General Settings, set locale to one with right-to-left directionality (e.g., Arabic / العربية )
- Check that sliders in, e.g., Player Settings, are right-to-left
- Check that switches/toggles in, e.g., Player Settings, are right-to-left
- Check that tooltips in, e.g., Player Settings, open correctly
- Check that arrow navigation works correctly by going forward one route, then going back (left = forward, right = back)
- Check that labels left-aligned in ltr languages are right-aligned (e.g., side-nav labels, video titles, search results, `ft-input` placeholders, comments sections, the labels in the profile sections)
- Check that rtl directionality and styling is maintained after Ctrl+R
- (Optional) Prepend `true ||` to `v-if` statements on lines 17 and 21 of `App.vue` to show & validate updates banner is rtl

## Desktop

- **OS:** OpenSUSE Tumbleweed
- **OS Version:** 2023xxxx
- **FreeTube version:** 0.19.0

## Future Action Items
### Add linter requiring use of directionality-neutral CSS properties
To maintain a usable right-to-left interface, we must ensure that the directionality-agnostic property names listed above are used in all cases going forward (with maybe the sole exception of `videoJS.css`).

### Ongoing Localization Problem
We currently use string concatenation of data and an i18n label in many places in the code (e.g., "v0.19.0 Beta" on the About page, "5,000 views" on video lists. This is not the way to do this, because many languages order those values differently. This will need to be addressed in a separate PR given that this is more of a localization problem, but I thought it would mention it specifically because languages with right-to-left directionality are most notably affected. 

### Enhancement: Switch left-to-right direction in `ft-input`s and their search results based on whether the text typed is in `ltr` or `rtl` language.

### To RTL language users
 let us know if you are having any issues with directionality in the video player or any other part of FreeTube!
